teps/0149-tekton-cli-local-data-upload.md

```diff
@@ -0,0 +1,225 @@
+
+Streaming data into a Kubernetes POD must allow configuration of the application reading the STDIN using [tar format][tarStandard] for a wide variety of workloads. When the configured executable is not found on the Step's container image the CLI returns error.
+
+The ability to upload local content must respect the Workspace definition, and the upload scope must be restricted to it.
+
+## Proposal
+
+Extend the `.spec.debug` attributes to allow developers to intervene on a `PipelineRun` or `TaskRun` workflow by declaring which Task or Step should be suspended. The workflow in pause can be subject to context inspection, execution of arbitrary commands and as well as leveraging the same machinery ["`kubectl cp`"][kubectlCPSnippet] uses to stream local data into the POD.
+
+This proposal is focused on enhancing the developer experience with the ability to work on an _inner-loop_, as in, testing local modifications against a Tekton defined CI as early as possible, as often as needed in order to iterate on the project at hand.
+
+Currently developers rely on pushing the local changes on a remote repository only to "_git clone_" them into a Pipeline, taking a considerable amount of time. This proposal pivots the path that allows a direct streaming of local changes into the live `PipelineRun`.
+
+To stream local data into a live `PipelineRun`/`TaskRun` requires the CLI to coordinate the execution workflow in order to run the command which will store the data coming via `STDIN`, most of which is already taking place in existing CLI features. The data streaming is a specialization of the debug abilities described in this document.
+
+Additionally, being able to interactively debug a Pipeline or Task allows developers to narrow their focus on a specific part, accessing the runtime environment, trying out scripting changes, asserting the results, rinse-and-repeat until the desired state. Another type of _inner-loop_ activity but for Tekton resource maintainers.
+
+The suspended step will interactively execute the designed command and script blocks using newly added "`debug-run`" script, showing what's being executed on the container and returning the control back to the user to either "`debug-continue`" or "`debug-fail-continue`".
+
+## Design Details
+
+### API
+
+Tekton's `PipelineRun` and `TaskRun` CRDs will be modified, the attribute "`.spec.debug.breakpoint`" will support "`onSteps`" and "`onTasks`", and additional attributes are introduced under "`.spec.debug`" listing a set of Task and Step names.
+
+When the resource lifecycle reaches one of the `.spec.debug.{steps,tasks}` the workflow is paused, allowing the user to take control. The objective of these attributes is to determine where the workflow will be paused.
+
+The following snippet shows the `TaskRun` modifications:
+
+```yml
+---
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata: {}
+spec:
+  debug:
+    breakpoint: ["onSteps"]
+    steps: ["step-name"]
```

Comment:
The assumption being that it's only possible to break on named steps, or do we want to provide an index option too?

Reply:
The slice index sounds like a good idea have, thanks for the suggestion. 

Yes, I think we should have the index. Then, should we have both named steps and index, or only one?

Thinking out loud, the index is less volatile than a step name, so perhaps is a better way to describe which step the debugger should stop at. 🤔